graphics/src/ColladaLoader.cc

```diff
@@ -1559,38 +1653,121 @@ void ColladaLoader::Implementation::LoadNormals(const std::string &_id,
+    totCount = std::stoi(floatArrayXml->Attribute("count"));
+  else
+  {
+    gzerr << "<float_array> has no count attribute in normal coordinate "
+          << "element with id[" << _id << "]\n";
+    return;
+  }
+
+  normalsXml = normalsXml->FirstChildElement("technique_common");
+  if (!normalsXml)
+  {
+    gzerr << "Unable to find technique_common element for normals "
+          << "coordinates with id[" << _id << "]\n";
+    return;
+  }
+
+  // Get the accessor XML element.
+  normalsXml = normalsXml->FirstChildElement("accessor");
+  if (!normalsXml)
+  {
+    gzerr << "Unable to find <accessor> as a child of <technique_common> "
+          << "for normals coordinates with id[" << _id << "]\n";
+    return;
+  }
+
+  // Read in the stride for the normals coordinate values. The stride
+  // indicates the number of values in the float array the comprise
+  // a complete normal coordinate.
+  if (normalsXml->Attribute("stride"))
+  {
+    stride = std::stoi(normalsXml->Attribute("stride"));
+  }
+  else
+  {
+    gzerr << "<accessor> has no stride attribute in normal coordinate "
+          << "element with id[" << _id << "]\n";
+    return;
+  }
+
+  // Nothing to read. Don't print a warning because the collada file is
+  // correct.
+  if (totCount == 0)
+    return;
 
   std::unordered_map<gz::math::Vector3d,
       unsigned int, Vector3Hash> unique;
 
   std::string valueStr = floatArrayXml->GetText();
-  std::istringstream iss(valueStr);
-  do
+  // std::istringstream iss(valueStr);
+
+  auto toDoubleVec = [](std::string_view sv, size_t totalCount)
   {
-    gz::math::Vector3d vec;
-    iss >> vec.X() >> vec.Y() >> vec.Z();
-    if (iss)
+    std::vector<double> result;
+    // Preallocate memory based on known count
+    result.reserve(totalCount * 3);
+    const char *start = sv.data();
+    char *end{};
+    while (true)
     {
-      vec = rotMat * vec;
-      vec.Normalize();
-      _values.push_back(vec);
+      double d = std::strtod(start, &end);
+      if (start == end)
+        break;
+      start = end;
+      if (errno == ERANGE)
+        throw std::runtime_error("strtod() overflow");
+      result.emplace_back(d);
+    }
+    return result;
+  };
+
+  auto values = toDoubleVec(valueStr, totCount);
+
+  gz::math::Vector3d vec;
+  std::shared_ptr<gz::math::Vector3d> prev_vec =
+    std::make_shared<gz::math::Vector3d>(gz::math::Vector3d::Zero);
+  if (!_values)
+    _values = std::make_shared<std::vector<gz::math::Vector3d>>();
+  if (!_duplicates)
+        _duplicates = std::make_shared<std::unordered_map<unsigned int,
+                                                          unsigned int>>();
+
+  for (int i = 0; i < totCount; i += stride)
+  {
+    vec.Set(values[i],
+            values[i+1],
+            values[i+2]);
+
+    vec = rotMat * vec;
+    vec.Normalize();
+    (*_values).emplace_back(vec);
 
-      // create a map of duplicate indices
+    // create a map of duplicate indices
+    if((*prev_vec) != vec){
```

Comment:
Looked into the test failure and found that this check here results in more unique normals. Not sure if I understand the logic here. A vector is marked as unique if it's not equal to the previous one? So say if we have the following vectors: A = [0, 0, 1], B= [0, 1, 0], C =[0, 0, 1]. I think with this logic, C will be marked as unique because it does not equal to B? 

If I remove the extra check with the following patch, the test passes:

<details><summary>collada.patch</summary>

```diff
diff --git a/graphics/src/ColladaLoader.cc b/graphics/src/ColladaLoader.cc
index b7979a3..c6c231c 100644
--- a/graphics/src/ColladaLoader.cc
+++ b/graphics/src/ColladaLoader.cc
@@ -1562,8 +1562,6 @@ void ColladaLoader::Implementation::LoadPositions(const std::string &_id,
   auto values = toDoubleVec(valueStr, totCount);
 
   gz::math::Vector3d vec;
-  std::shared_ptr<gz::math::Vector3d> prev_vec =
-    std::make_shared<gz::math::Vector3d>(gz::math::Vector3d::Zero);
   if (!_values)
     _values = std::make_shared<std::vector<gz::math::Vector3d>>();
   if (!_duplicates)
@@ -1579,16 +1577,10 @@ void ColladaLoader::Implementation::LoadPositions(const std::string &_id,
     (*_values).emplace_back(vec);
 
     // create a map of duplicate indices
-    if((*prev_vec) != vec){
-      if (unique.find(vec) != unique.end())
-        (*_duplicates)[(*_values).size()-1] = unique[vec];
-      else
-        unique[vec] = (*_values).size()-1;
-    }
+    if (unique.find(vec) != unique.end())
+      (*_duplicates)[(*_values).size()-1] = unique[vec];
     else
       unique[vec] = (*_values).size()-1;
-
-    (*prev_vec) = vec;
   }
 
   this->positionDuplicateMap[_id] = _duplicates;
@@ -1728,8 +1720,6 @@ void ColladaLoader::Implementation::LoadNormals(const std::string &_id,
   auto values = toDoubleVec(valueStr, totCount);
 
   gz::math::Vector3d vec;
-  std::shared_ptr<gz::math::Vector3d> prev_vec =
-    std::make_shared<gz::math::Vector3d>(gz::math::Vector3d::Zero);
   if (!_values)
     _values = std::make_shared<std::vector<gz::math::Vector3d>>();
   if (!_duplicates)
@@ -1747,16 +1737,10 @@ void ColladaLoader::Implementation::LoadNormals(const std::string &_id,
     (*_values).emplace_back(vec);
 
     // create a map of duplicate indices
-    if((*prev_vec) != vec){
-      if (unique.find(vec) != unique.end())
-        (*_duplicates)[(*_values).size()-1] = unique[vec];
-      else
-        unique[vec] = (*_values).size()-1;
-    }
+    if (unique.find(vec) != unique.end())
+      (*_duplicates)[(*_values).size()-1] = unique[vec];
     else
       unique[vec] = (*_values).size()-1;
-
-    (*prev_vec) = vec;
   }
 
   this->normalDuplicateMap[_id] = _duplicates;
@@ -1917,8 +1901,6 @@ void ColladaLoader::Implementation::LoadTexCoords(const std::string &_id,
   auto values = toDoubleVec(valueStr, totCount);
 
   gz::math::Vector2d vec;
-  std::shared_ptr<gz::math::Vector2d> prev_vec =
-    std::make_shared<gz::math::Vector2d>(gz::math::Vector2d::Zero);
   if (!_values)
     _values = std::make_shared<std::vector<gz::math::Vector2d>>();
   if (!_duplicates)
@@ -1933,16 +1915,10 @@ void ColladaLoader::Implementation::LoadTexCoords(const std::string &_id,
     (*_values).emplace_back(vec);
 
     // create a map of duplicate indices
-    if((*prev_vec) != vec){
-      if (unique.find(vec) != unique.end())
-        (*_duplicates)[(*_values).size()-1] = unique[vec];
-      else
-        unique[vec] = (*_values).size()-1;
-    }
+    if (unique.find(vec) != unique.end())
+      (*_duplicates)[(*_values).size()-1] = unique[vec];
     else
       unique[vec] = (*_values).size()-1;
-
-   (*prev_vec) = vec;
   }
 
   this->texcoordDuplicateMap[_id] = _duplicates;
```

</details>